### ISSUE TITLE
Nightly regression: False-positive `dead_code` warning for fields used in Debug output

### ISSUE THREAD

**Issue by chanks** [rust-lang/rust#89721]
I tried this code:

```rust
#[derive(Debug)]
enum MyError {
    MyVariant { more_info: &'static str },
}

fn main() {
    my_function().unwrap()
}

fn my_function() -> Result<(), MyError> {
    Err(MyError::MyVariant { more_info: "hello" })
}
```

I expected to see this happen: No dead code warning, since `more_info` is emitted in the debug output when panicking. This is, in fact, the behavior on current stable.

Instead, this happened (on current nightly):
```
warning: field is never read: `more_info`
 --> src/main.rs:3:17
  |
3 |     MyVariant { more_info: &'static str },
  |                 ^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default
```

Reproduction on Rust Playground: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=a0c9dd25de81250cead3d61a16f99497

### Meta

`rustc --version --verbose`:
```
rustc 1.57.0-nightly (a8f2463c6 2021-10-09)
binary: rustc
commit-hash: a8f2463c68a6532d74a13ec402ec5b513e4e2726
commit-date: 2021-10-09
host: x86_64-unknown-linux-gnu
release: 1.57.0-nightly
LLVM version: 13.0.0
```

The bug does not appear on current stable, as you can see by toggling the compiler version in the Rust Playground link above.

Possibly related: https://github.com/rust-lang/rust/issues/89717 (the code triggering the issue is somewhat different, but it's another dead_code false positive that just showed up on nightly).

Thank you!

**Comment by ehuss** [comment]
Thanks for the report!  I'm going to close this as a duplicate of #88900, as it is essentially the same issue.  The `dead_code` lint was extended to ignore Debug impls, as placing `derive(Debug)` on any struct resulted in never displaying a warning for unused fields.
